### DISCUSSION TITLE
Displaying IPython.core.display.HTML object  in Quarto

### DISCUSSION THREAD

**Discussion by flitaiff** [quarto-dev/quarto-cli#6051] [Q&A]
Hi there,

Any help would be much appreciated I can´t display a HTML object or even a HTML file. The rendered document appears with nothing in it. Maybe I am forgetting something basic here, bit I am stuck

Environment: RStudio with a  Quarto document  running Python 3.11.4, downloaded from the official site, All packages installed.
The files for the upload are attached. You can reproduce it;

Header:
````qmd
---
title: "LDA_GENSIM"
author: "Scott Free"
editor: visual
jupyter: python3
format:
  html:
    embed-resources: true
    self-contained: true
execute:
  cache: true
---

## LDA with GENSIM

```{python}
#| label: "load-packages"
#| include: false
#| echo: false


import gensim
import tomotopy as tp
import pyLDAvis
import IPython
import warnings
from pyLDAvis import urls, prepared_data_to_html, gensim_models
from IPython.display import display, HTML
from gensim.models import LdaModel
from gensim.corpora import Dictionary, MmCorpus
from IPython.display import display


```


```{python}
#| label: dtmt_gensim_viztemp
#| include: false
#| eval: true
#| dependson: "load-packages"
#| format:
#|   html:
#|     self-contained: true
#|     embed-resources: true


# Load the LDA model, dictionary, and corpus for all documents
lda_model = LdaModel.load('lda_model_all.gensim')
dictionary = Dictionary.load('dictionary_all.gensim')
corpus = MmCorpus('corpus_all.mm')

# Prepare the visualization data
vis_data = gensim_models.prepare(lda_model, corpus, dictionary, n_jobs=1)


# This should display a pretty comprehensive interactive 
# graphs on LDA Topic Models.

vis_html = pyLDAvis.display(vis_data) 

display(vis_html)


# If doesn´t work, lest´s try working with the HTML object

# Forcing the render of the HTML
display(HTML(vis_html.data))

# If it doesn´t work either, lets try with rendering the raw HTML 
#    by saving it in a file. If you check it, it exists with the graph needed
pyLDAvis.save_html(vis_data, 'lda_visualization.html')

# and then, manually rending it
display(HTML('lda_visualization.html'))

# Nothing happens. In every try the message bellow appears in the standard output:
# >>> <IPython.core.display.HTML object>
# And that´s it.

```


````

By the way, the HTML file is properly created in disk. Nothing is displayed.

Any ideas? Am I forgetting something?

Here are the files needed to reproduce the problem.

[FILES - 3.zip](https://github.com/quarto-dev/quarto-cli/files/11898304/FILES.-.3.zip)

**Comment by cscheid** [comment]
Can you provide a self-contained example that we can attempt to reproduce? (For example, adding the files and .qmd document to a github repository)

**Reply by flitaiff** [reply]
Hi, I just have updated the script and uploaded the files needed.

**Reply by cscheid** [reply]
Can you also include the full actual .qmd file that causes trouble for you? In the issue, your snippet is missing at least a closing triple-tick marker, and I'm wondering if there's more truncated output.

**Reply by mcanouil** [reply]
@flitaiff see my example in <https://github.com/quarto-dev/quarto-cli/discussions/6051#discussioncomment-6306129>

**Comment by mcanouil** [comment]
You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by mcanouil** [reply]
@flitaiff where si you find that document? It's odd that a document on Quarto website contains both `self-contained` and `embed-resources` ...

**Reply by flitaiff** [reply]
Hi, it´s probably wrong, but what would be the correct parameters? I think the Quarto documentation lacks more examples. RMarkdown has a much more mature knowledge base so far.

**Reply by mcanouil** [reply]
It is clearly stated that self-contained is deprecated in every pages of the documentation as far as I know. Since you said you took the example out of the documentation, could you tell us where for us to fix this inconsistency?
Also I am a bit concerned with the fact that you define format inside code cells, which even in rmarkdown is not valid.

Finally, could you make your example as simple as possible?
You currently have a lot of things that I don't think are relevant here.

**Reply by flitaiff** [reply]
I have updated the script, you can paste it and use the files as needed. Sorry, I should have been more clear in the first place.

**Reply by mcanouil** [reply]
Still, you did not answer my questions/comments.

**Reply by flitaiff** [reply]
The .qmd is inside the zip file I just sent,
This was suggested by Bing chat. I see now it is being deprecated, but I was trying anything I could to make it work. This is a huge bottleneck in my work right now.

**Reply by flitaiff** [reply]
By the way, I removed the format inside the code cells and removed the "self-contained" from the header.
Nothing changed.

**Reply by mcanouil** [reply]
Since I am away from a keyboard, I cannot be sure, but using display on a display element seems very wrong (as the double display import).
See <https://pyldavis.readthedocs.io/en/latest/modules/API.html#pyLDAvis.display>.

**Reply by flitaiff** [reply]
If you check the file generated, lda_visualization.html it should show what I am looking for, For all that matters, it looks pyLDAvis.display(vis data) is working properly. I just can´t render the Quarto document with it.

**Reply by mcanouil** [reply]
I tried your files, but your example is not reproducible as your saved model can't be loaded.

**Reply by flitaiff** [reply]
Did you have installed the right modules? It works here, just not rendered into the Quarto document. The HTML generated is below. 
[lda_visualization.zip](https://github.com/quarto-dev/quarto-cli/files/11905454/lda_visualization.zip)

**Reply by mcanouil** [reply]
Since I showed that the modules are working properly within a 100% reproducible repository, I am 100% sure your issue has nothing to do with Quarto.

I highly suggest you use my example and update it to fit your needs.

See <https://github.com/quarto-dev/quarto-cli/discussions/6051#discussioncomment-6308307>.

If you still have issues, I strongly suggest to ask on a Python oriented platform as your issues are with Python modules usage.

For instance and as I mentioned before, you imported twice `display()` while your model module also has `display()`, thus you are basically generating conflicts in your code.
Also, you should not use AI based code without knowing what's the code is doing.

**Reply by flitaiff** [reply]
Anyway, I just managed to show the HTML outside the code block with 
<iframe src="lda_visualization.html" width="100%" height="1000"></iframe>
Suboptimal, but worked.

**Reply by flitaiff** [reply]
The problem appears to be related to the inability to display an HTML file within the code block, as you can see with the code `HTML(vis_html.data)`. This issue is not associated with the model itself but likely comes from the interaction between RStudio and Quarto. I had hoped that observing the repeated appearance of '<IPython.core.display.HTML object>' (with no other output) in the RStudio console standard output when attempting to display anything would provide some insight.

I'm curious to know the specific error you encountered while loading the model, dictionary, and corpus before preparing and displaying them using `pyLDAvis.display`. In any case, here are the files once again:

[FILES - 3.zip](https://github.com/quarto-dev/quarto-cli/files/11910239/FILES.-.3.zip)

Regarding AI-generated code (or anything for that matter), I am fully aware that blindly accepting output from AI can have disastrous consequences. I tried it as a last resort before seeking assistance here.

Before switching back to RMarkdown, I will make an attempt to run the code in a Jupyter notebook and see if ir renders properly being called within the code block. I will let you know if I succeed.

**Reply by mcanouil** [reply]
My example still works perfectly when saving the visualisation to HTML and importing it back.
<img width="761" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/8af533e3-611c-48f7-9e45-70260cac321f">

See <https://github.com/mcanouil/quarto-issues/tree/main/quarto-cli-6051>.

**Comment by mcanouil** [comment]
See a fully reproducible and working example showing there is no issue using the Python module and Quarto.

Code: https://github.com/mcanouil/quarto-issues/tree/main/quarto-cli-6051
